File: crivo-de-eratostenes/crivo.cpp
#include <iostream>
#include <math.h>
#include <vector>
#include <chrono>

using namespace std;

int main() {
    // Espaço analisado.
    int valorLimite = 100000;

    // Raíz quadrada do valor limite analisado, determinando o maior número a ser analisado.
    int multiploLimite = floor(sqrt(valorLimite));

    // Considero que todos os números são primos, inicialmente.
    vector<bool> isPrime(valorLimite, true);

    // Removo os valores inicias (0 e 1), que sei que não são primos.
    isPrime[0] = false;
    isPrime[1] = false;

    auto tempoInicial = chrono::high_resolution_clock::now();

    // Começo a rodar o algoritmo do crivo à partir do primeiro primo conhecido (2).
    for (int valorAnalisado = 2; valorAnalisado < multiploLimite; valorAnalisado++) {

        // Se o valor analisado não é primo, eu pulo seu crivo.
        if (isPrime[valorAnalisado] == false) {
            continue;
        }

        // Se é primo, eu removo todos os próximos itens que são multiplos dele.
        for (int j = valorAnalisado+1; j < isPrime.size(); j++) {
            if (j % valorAnalisado == 0) {
                isPrime[j] = false;
            }
        }
    }

    auto tempoFinal = chrono::high_resolution_clock::now();

    int qtdPrimos = 0;

    // Imprimindo a lista de números primos e contando seu total.
    for (int i = 0; i < isPrime.size(); i++) {
        if (isPrime[i] == true) {
            cout << i << endl;
            qtdPrimos++;
        }
    }

    cout << "# de primos encontrados: " << qtdPrimos << endl;
    cout << "tempo: " << (std::chrono::duration_cast<std::chrono::nanoseconds>(tempoFinal - tempoInicial).count()) / 1000000.0 << " ms" << endl;

    return 0;
}

File: crivo-de-eratostenes/Crivo.java
public class Crivo {

    public static void main(String[] args) {
        
        // Espaço analisado.
        int valorLimite = 10000000;

        // Raíz quadrada do valor limite analisado, determinando o maior número a ser analisado.
        int multiploLimite = (int) Math.floor(Math.sqrt(valorLimite));

        boolean[] isPrime = new boolean[valorLimite];

        // Considero que todos os números são primos, inicialmente.
        for (int i = 0; i < isPrime.length; i++) {
            isPrime[i] = true;
        }

        long tempoInicial = System.nanoTime();

        // Otimização: Removo números que eu sei que não são primos:
        // 0, 1 e todos os pares que não sejam o 2.
        isPrime[0] = false;
        isPrime[1] = false;
        for (int i = 4; i < isPrime.length; i+=2) {
            isPrime[i] = false;
        }

        // Começo a rodar o algoritmo do crivo à partir do primeiro primo ímpar conhecido (3).
        for (int valorAnalisado = 3; valorAnalisado < multiploLimite; valorAnalisado+=2) {

            // Se o valor analisado não é primo, eu pulo seu crivo.
            if (isPrime[valorAnalisado] == false) {
                continue;
            }

            // Se é primo, eu removo todos os próximos itens ímpares que são multiplos dele.
            for (int j = valorAnalisado+2; j < isPrime.length; j+=2) {
                if (j % valorAnalisado == 0) {
                    isPrime[j] = false;
                }
            }
        }

        long tempoFinal = System.nanoTime();

        int qtdPrimos = 0;

        // Imprimindo a lista de números primos e contando seu total.
        for (int i = 0; i < isPrime.length; i++) {
            if (isPrime[i] == true) {
                System.out.println(i);
                qtdPrimos++;
            }
        }

        System.out.println("# de primos encontrados: " + qtdPrimos);
        System.out.println("tempo: " + (tempoFinal - tempoInicial) / 1_000_000.0 + " ms");
    }
}